common/lib/capa/capa/responsetypes.py

```diff
@@ -1351,21 +1492,114 @@ def setup_response(self):
+                            cmap.set(aid, 'incorrect')
+
             answer_variable = self.get_student_answer_variable_name(student_answers, aid)
             if answer_variable:
                 cmap.set_property(aid, 'answervariable', answer_variable)
+
         return cmap
 
     def get_answers(self):
-        amap = dict([(af.get('id'), contextualize_text(af.get(
+        """
+        Returns a dictionary with problem ids as keys.
+        Each entry is a list of the correct answers for that id.
+        """
+        answer_map = dict([(af.get('id'), contextualize_text(af.get(
             'correct'), self.context)) for af in self.answer_fields])
-        return amap
+
+        # Get the list of correct answers.
+        # Split to allow multiple correct answers and trim whitespace.
+        for aid in answer_map:
+            answer_map[aid] = answer_map[aid].split(',')
+            for index, word in enumerate(answer_map[aid]):
+                answer_map[aid][index] = word.strip()
+
+        # log.debug('%s: expected answers=%s' % (unicode(self),answer_map))
+        return answer_map
+
+    def get_partial(self):
+        """
+        Returns a dictionary similar to the answer map,
+        but with all the partially-correct answers instead.
+        """
+        partial_map = dict([(af.get('id'), contextualize_text(af.get(
+            'partial'), self.context)) for af in self.answer_fields])
+
+        # Get the list of partially-correct answers.
+        # Split to allow multiple options and trim whitespace.
+        for aid in partial_map:
+            if partial_map[aid] is not None:
+                partial_map[aid] = partial_map[aid].split(',')
+                for index, word in enumerate(partial_map[aid]):
+                    partial_map[aid][index] = word.strip()
+
+        # log.debug('%s: partially correct answers=%s' % (unicode(self),answer_map))
+        return partial_map
+
+    def get_partial_points(self, partial_map):
+        """
+        This dictionary matches the one returned by get_partial,
+        but gives the point value for the responses instead.
+        The default is 50% credit.
+        """
+
+        default_credit = 0.5
+
+        points_map = dict([(af.get('id'), contextualize_text(af.get(
+            'point_values', default=None), self.context)) for af in self.answer_fields])
+
+        # Get the partial credit for each partially-correct option.
+        # Split to allow multiple options and trim whitespace.
+        for aid in points_map:
+            if points_map[aid] is not None:
+                points_map[aid] = points_map[aid].split(',')
+                for index, word in enumerate(points_map[aid]):
+                    points_map[aid][index] = float(word.strip())
+            else:
+                points_map[aid] = [default_credit] * len(partial_map[aid])
+        # log.debug('%s: partial point values=%s' % (unicode(self),answer_map))
+        return points_map
```

Comment:
There are three cut-and-paste functions here, with only minor variations (most of which are not necessary). DRY. There are a number of ways to do this which are all okay. As a matter of personal style, I would return a single data structure with all three, and use that, but it'd also be okay to e.g. have a get_problem_map style function which takes either 'point_values', 'partial', or 'correct' as one parameter, and default as a second, or a number of other approaches to avoid the cut-and-paste.